src/Covenant/ASG.hs

```diff
@@ -897,49 +906,93 @@ cata rAlg rVal =
+                unless (Vector.length bfTyArgs > 0) (throwError . CataNotAnAlgebra $ t)
+                let lastTyArg = Vector.last bfTyArgs
+                unless (nev NonEmpty.! 1 == lastTyArg) (throwError . CataNotAnAlgebra $ t)
+                appliedArgT <- case valT of
+                  BuiltinFlat bT -> case bT of
+                    ByteStringT -> do
+                      unless (bfName == "ByteString_F") (throwError . CataUnsuitable algT $ valT)
+                      pure $ Datatype "ByteString_F" . Vector.singleton $ lastTyArg
+                    IntegerT -> do
+                      let isSuitableBaseFunctor = bfName == "Natural_F" || bfName == "Negative_F"
+                      unless isSuitableBaseFunctor (throwError . CataUnsuitable algT $ valT)
+                      pure $ Datatype bfName . Vector.singleton $ lastTyArg
+                    _ -> throwError . CataWrongBuiltinType $ bT
+                  Datatype tyName tyVars -> do
+                    lookedUp <- asks (view (#datatypeInfo % at tyName))
+                    case lookedUp of
+                      Nothing -> throwError . CataNoSuchType $ tyName
+                      Just info -> case view #baseFunctor info of
+                        Just (DataDeclaration actualBfName _ _ _, _) -> do
+                          unless (bfName == actualBfName) (throwError . CataUnsuitable algT $ valT)
+                          let lastTyArg' = stepDownDB lastTyArg
+                          pure . Datatype bfName . Vector.snoc tyVars $ lastTyArg'
+                        _ -> throwError . CataNoBaseFunctorForType $ tyName
+                  _ -> throwError . CataWrongValT $ valT
+                resultT <- tryApply algT appliedArgT
+                refTo . AValNode resultT . CataInternal rAlg $ rVal
+              _ -> throwError . CataNotAnAlgebra $ t
+          _ -> throwError . CataNonRigidAlgebra $ algT
         t -> throwError . CataNotAnAlgebra $ t
     t -> throwError . CataApplyToNonValT $ t
 
 -- Helpers
 
+-- Note (Koz, 13/08/2025): We need this procedure specifically for `cata`. The
+-- reason for this has to do with how we construct the 'base functor form' of
+-- the value to be torn down by the catamorphism, in order to use the
+-- unification machinery to get the type of the final result.
+--
+-- To be specific, suppose we have `<List_F r (Maybe r) -> !Maybe r>` as our algebra
+-- argument (where `r` is some rigid), and `List r` as the value to be torn
+-- down. If we assume the rigid is bound one scope away, `r`'s DeBruijn index
+-- will be `S Z` for
+-- the value to be torn down, but `S (S Z)` for the algebra argument. The way
+-- our approach works is:
+--
+-- 1. Look at the algebra argument, specifically the base functor type. Take its
+--    last type argument, which we will call `last`.
+-- 2. Determine the base functor for the value to be torn down. Cook up a new
+--    instance of the base functor type, copying all the type arguments from the
```

Comment:
This isn't essential, but it might be nice to have an example for this process. I *think* I understand what we're doing / why we're doing it, though it would be easier to follow if each of the steps referred to the motivating example you give on line 961.